pkg/gadgets/profile/tracer/core/tracer.go

```diff
@@ -0,0 +1,380 @@
+		// 0000000000000000 t acpi_video_unregister_backlight      [video]
+		// First is the symbol address and second is described in man nm.
+		symbols = append(symbols, kernelSymbol{
+			addr: addr,
+			name: fields[2],
+		})
+	}
+
+	err = scanner.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return symbols, nil
+}
+
+// findKernelSymbol tries to find the kernel symbol corresponding to the given
+// instruction pointer.
+// For example, is instruction pointer is 0x1004 and there is a symbol which
+// address is 0x1000, this function will return the name of this symbol.
+// If no symbol is found, it returns "[unknown]".
+func findKernelSymbol(kAllSyms []kernelSymbol, ip uint64) string {
+	// Go translation of iovisor/bcc ksyms__map_addr():
+	// https://github.com/iovisor/bcc/blob/c65446b765c9f7df7e357ee9343192de8419234a/libbpf-tools/trace_helpers.c#L149
+	end := len(kAllSyms) - 1
+	var addr uint64
+	start := 0
+
+	for start < end {
+		mid := start + (end-start+1)/2
+
+		addr = kAllSyms[mid].addr
+
+		if addr <= ip {
+			start = mid
+		} else {
+			end = mid - 1
+		}
+	}
+
+	if start == end && kAllSyms[start].addr <= addr {
+		return kAllSyms[start].name
+	}
+
+	return "[unknown]"
+}
+
+func getReport(t *Tracer, kAllSyms []kernelSymbol, stack *ebpf.Map, keyValue keyValue) (types.Report, error) {
+	kernelInstructionPointers := [perfMaxStackDepth]uint64{}
+	userInstructionPointers := [perfMaxStackDepth]uint64{}
+	v := keyValue.value
+	k := keyValue.key
+
+	// 	if (!env.kernel_stacks_only && k->user_stack_id >= 0) {
+	if k.user_stack_id >= 0 {
+		err := stack.Lookup(k.user_stack_id, unsafe.Pointer(&userInstructionPointers))
+		if err != nil {
+			return types.Report{}, err
+		}
+	}
+
+	// 	if (!env.user_stacks_only && k->kern_stack_id >= 0) {
+	if k.kern_stack_id >= 0 {
+		err := stack.Lookup(k.kern_stack_id, unsafe.Pointer(&kernelInstructionPointers))
+		if err != nil {
+			return types.Report{}, err
+		}
+	}
+
+	userSymbols := []string{}
+	for _, ip := range userInstructionPointers {
+		if ip == 0 {
+			break
+		}
+
+		// We will not support getting userland symbols.
+		userSymbols = append(userSymbols, "[unknown]")
+	}
+
+	kernelSymbols := []string{}
+	for _, ip := range kernelInstructionPointers {
+		if ip == 0 {
+			break
+		}
+
+		kernelSymbols = append(kernelSymbols, findKernelSymbol(kAllSyms, ip))
+	}
+
+	report := types.Report{
+		Node:        t.node,
+		Comm:        C.GoString(&k.name[0]),
+		Pid:         uint32(k.pid),
+		UserStack:   userSymbols,
+		KernelStack: kernelSymbols,
+		Value:       v,
+	}
+
+	container := t.resolver.LookupContainerByMntns(uint64(k.mntns_id))
+	if container != nil {
+		report.Container = container.Name
+		report.Pod = container.Podname
+		report.Namespace = container.Namespace
+	}
+
+	return report, nil
+}
+
+func (t *Tracer) Stop() (string, error) {
+	reports := []types.Report{}
+
+	keysValues, err := readCountsMap(t.objs.profileMaps.Counts)
```

Comment:
There should be a check that `t.objs.profileMaps.Counts` to avoid crashing if the map hasn't been created when this function is called.